Quilt 支持

Pigeon0v0 opened the pull request:
与 NeoForge 类似，本 PR 将基于先前的 Fabric 安装流程，加入针对 Quilt 的支持。
Close #48 

<details>
<summary>效果预览 / Preview</summary>

![Quilt-1](https://github.com/Hex-Dragon/PCL2/assets/60414767/05f9b092-0799-47a9-8fbf-cd227d3e90aa)

![Quilt-2](https://github.com/Hex-Dragon/PCL2/assets/60414767/0a82dec2-5f97-4d94-a7ce-d05e96f0686b)

![Quilt-3](https://github.com/Hex-Dragon/PCL2/assets/60414767/e216f52e-3fb2-4190-82ca-237932e7e96e)

![Quilt-4](https://github.com/Hex-Dragon/PCL2/assets/60414767/4cd05dc3-a3ca-4f3f-9578-1405c72befa1)
</details>


To-Do List: 

- [x] 在自动安装界面加入 Quilt 选项，并支持显示版本列表
- [x] Quilt 相关信息获取
- [x] 版本列表相关处理（版本排序，Quilt 相关信息显示，附加安装选项之间冲突等）
- [x] Quilt 自动安装
- [x] Quilt 手动安装界面
- [x] 相关资源文件（图标等）补充
- [x] Quilt 整合包安装
- [x] 其他界面选项和文本补充

Known Issues: 
- 手动安装包图标需要补充

**Note:** 由于 Quilt 同步 API 的问题，BMCLAPI 暂时无法用于下载 Quilt。

~先前的 PR 操作不当关了（）~

Pigeon0v0 committed:
refactor: 重新整了下 Quilt，但是还没完工

Pigeon0v0 committed:
Merge pull request #5 from Pigeon0v0/main

refactor: 重新整了下 Quilt，但是还没完工

Pigeon0v0 committed:
feat: Quilt 支持（未完工）

Pigeon0v0 committed:
feat: Quilt 支持

Pigeon0v0 committed:
feat: Quilt 整合包安装

Pigeon0v0 commented:
@3gf8jv4dv 

Please have a test on features related, such as modpack installing, auto installing, etc. Appreciate!

3gf8jv4dv reviewed (changes requested):
1. From the API results, the lowest version supported by Quilt is only 1.14.4, but 1.14~1.14.3 can still choose Quilt. Of course, the installation is definitely unsuccessful. It is recommended that the Quilt option not be displayed for versions prior to 1.14.4.

2. In *Select Instance (版本选择)*, Quilt instances are categorized as *Fabric Instances (Fabric 版本)*

   <details>
   <summary>Preview</summary>

   ![](https://github.com/Hex-Dragon/PCL2/assets/158490944/89672b21-450f-4ddf-84da-e22daf13c0e4)

   </details>

3. Some QFAPI / QSL that are marked as supporting multiple versions (only on one platform) may not match. For example, when installing Quilt with 1.20, after selecting Quilt Loader, there are only two results in the QFAPI / QSL list. However, after filtering the corresponding projects of CurseForge / Modrinth with the *1.20* version condition, there are more results (most of which are marked as supporting 1.20~1.20.1). About `QFAPI 2.0.0-beta.1 and QSL 2.0.0-beta.8` and `QFAPI 2.0.0-beta.3 and QSL 2.0.0-beta.9`, Modrinth is marked as supporting multiple versions, but CurseForge only shows one of them, and the other shows only supporting 1.19. QFAPI / QSL are shown as empty in these game versions.

   <details>
   <summary>Preview</summary>

   ![](https://github.com/Hex-Dragon/PCL2/assets/158490944/0a15e391-9d20-4104-8397-042ca4e0bd2a)

   </details>

4. For QFAPI / QSL, the current behavior is to only display *QSL*. Strictly speaking QFAPI and QSL are different things (there is a reason why the Quilt team specifically writes this). My opinion is that you can just write *qsl* for the relevant markup within the code (like in the links on the [CurseForge](https://www.curseforge.com/minecraft/mc-mods/qsl) / [Modrinth](https://modrinth.com/mod/qsl) platform), but write *QFAPI / QSL* for the text displayed to the user.

5. Some Quilt Loader versions with `-pre` are marked as *stable (稳定版)*.

*Tried some Quilt modpacks on Modrinth and seems to be OK.*

SALTWOOD commented:
![image](https://github.com/Hex-Dragon/PCL2/assets/105980161/c0fece29-8dbc-4c78-8edb-0ce946d634ec)
无法选中Quilt的选项……也许是个bug？

SALTWOOD commented:
![image](https://github.com/Hex-Dragon/PCL2/assets/105980161/4595e479-874a-4ee9-a3df-6e8741e8dc6d)
手动安装出现错误

Pigeon0v0 committed:
refactor: 重构 Mod Loader 冲突判断
fix: 疑似一些正则错误

SALTWOOD commented:
> @SALTWOOD
> 
> > 无法选中Quilt的选项……也许是个bug？
> 
> I am curious about how you got triggered (I've only encountered it once, and it was inexplicable at the time, I do not know how it was triggered).
> 
> > 手动安装出现错误
> 
> > * [x]  Quilt 手动安装界面
> 
> From the author’s description, it seems to be in line with expectations, but problems do exist.
> 
> @Pigeon0v0
> 
> After writing four opinions, I suddenly found a bug, so I added it to the previous review.

I don't know. Every time I run it, the bug appears.

SALTWOOD commented:
也许是因为 Pages\PageDownload\PageDownloadInstall.xaml.vb Line 347-560没加？

SALTWOOD commented:
I turned on the catch for the `InvalidCastException` and found that this exception was thrown here:
![image](https://github.com/Hex-Dragon/PCL2/assets/105980161/9024aaad-64eb-472a-92a6-d5785d8d3b1f)
System.InvalidCastException
  HResult=0x80004002
  Message=无法将类型为“Newtonsoft.Json.Linq.JArray”的对象强制转换为类型“Newtonsoft.Json.Linq.JObject”。
  Source=Plain Craft Launcher 2
  StackTrace:
   在 PCL.ModDownload.DlQuiltListOfficialMain(LoaderTask`2 Loader) 在 C:\Users\SALTWOOD\Desktop\p\Plain Craft Launcher 2\Modules\Minecraft\ModDownload.vb 中: 第 1132 行

SALTWOOD commented:
both 两个bug在我这边都能稳定地复现

SALTWOOD commented:
@3gf8jv4dv Can you tell me what the return value of this API is like on your side?
`https://[meta.quiltmc.org/v3/versions/game](https://meta.quiltmc.org/v3/versions/game)`

Pigeon0v0 commented:
> ![image](https://github.com/Hex-Dragon/PCL2/assets/105980161/4595e479-874a-4ee9-a3df-6e8741e8dc6d)
> 
> 手动安装出现错误

这个 BUG 我记得我修了的...

SALTWOOD commented:
草，没切分支（

SALTWOOD commented:
我在想一个东西……就是说一部分模组在使用QFAPI加QSL的搭配的时候会无法使用，于是就会在QUILT下使用FABRIC API
而理论上quilt作为fabric的一个分支，当然是可以使用fabric API的
所以有没有必要也把fabric API的安装选项显示出来呢

SALTWOOD commented:
并且在同时选择了quilt和optifine之后……

Pigeon0v0 commented:
需要确认一下 OptiFabric 和 Quilt 的兼容情况

SALTWOOD commented:
那么就只能是先暂时禁用

SALTWOOD commented:
并且在版本选择里头安装了quilt的版本还是被识别为fabric

SALTWOOD commented:
> > 并且在版本选择里头安装了quilt的版本还是被识别为fabric
> 
> I mentioned this before.

all right

Pigeon0v0 commented on Plain Craft Launcher 2/Resources/Custom.xaml:
```diff
@@ -206,6 +206,8 @@
 			<Image Width="30" Margin="5,0,0,0" Source="pack://application:,,,/images/Blocks/Egg.png"/>
 			<!-- 布料（Fabric 图标） -->
 			<Image Width="30" Margin="5,0,0,0" Source="pack://application:,,,/images/Blocks/Fabric.png"/>
+      <!-- 方格（Quilt 图标） -->
+      <Image Width="30" Margin="5,0,0,0" Source="pack://application:,,,/images/Blocks/Quilt.png"/>
```
This should have no problem.

Pigeon0v0 committed:
fix: 正则表达式错误
fix: QFAPI / QSL 版本适配判断可能有误
fix: Quilt 核心判断和卡片分类错误；Quilt 测试版可能误判为稳定版
fix: 来自 Curse 的 Quilt 整合包安装可能失败
opti: 当 MC 版本小于 1.14.4 时不显示 Quilt 自动安装选项
fix: 先选择一个 Mod Loader 后直接返回 MC 版本选择，再选择另一个 MC 版本时不会重置附加安装项状态
fix: 默认的核心命名可能导致异常
opti: 不允许 Quilt 与 OptiFine 同时安装
chores: QSL -> QFAPI / QSL
chores: 去掉了 Quilt 官网的 "www."

Pigeon0v0 commented:
d02adaf should solve issues mentioned above. Have a try and leave any comments! :D

@3gf8jv4dv @SALTWOOD

Pigeon0v0 committed:
chores: json -> Json; jar -> Jar

Pigeon0v0 commented:
3a1920c 顺手改了一些大小写，至少现在显示的都是 Json 和 Jar 了...

关于其他大小写和中英文空格之类的修改，日后还会再开一个 PR 之类的讨论一下...

Pigeon0v0 committed:
chores: 整合包安装时部分错误提示可能有误

3gf8jv4dv reviewed (commented):
No other issues encountered so far. Waiting for other people’s opinions.

LTCatt commented on Plain Craft Launcher 2/Modules/Minecraft/ModDownload.vb:
```diff
@@ -1075,6 +1075,95 @@
+    ''' </summary>
+    Public DlQuiltListOfficialLoader As New LoaderTask(Of Integer, DlQuiltListResult)("DlQuiltList Official", AddressOf DlQuiltListOfficialMain)
+    Private Sub DlQuiltListOfficialMain(Loader As LoaderTask(Of Integer, DlQuiltListResult))
+        Dim GameResult As JArray = NetGetCodeByRequestRetry("https://meta.quiltmc.org/v3/versions/game", IsJson:=True)
+        Dim LoaderResult As JArray = NetGetCodeByRequestRetry("https://meta.quiltmc.org/v3/versions/loader", IsJson:=True)
+        Dim InstallerResult As JArray = NetGetCodeByRequestRetry("https://meta.quiltmc.org/v3/versions/installer", IsJson:=True)
```
建议使用 `https://meta.quiltmc.org/v3/versions`，只需要一次请求

LTCatt commented on Plain Craft Launcher 2/Modules/Minecraft/ModMinecraft.vb:
```diff
@@ -382,18 +382,23 @@ Public Module ModMinecraft
-                        Regex = RegexSeek(LibrariesString, "(?<=((fabricmc)|(quiltmc)):intermediary:)[^""]*")
+                        Regex = RegexSeek(LibrariesString, "(?<=(fabricmc):intermediary:)[^""]*")
                         If Regex IsNot Nothing Then
                             _Version.McName = Regex
                             GoTo VersionSearchFinish
                         End If
-                        'FUTURE: [Quilt 支持] 从 Quilt 版本中获取版本号
-                        '从 jar 项中获取版本号
+                        '从 Quilt 版本中获取版本号
+                        Regex = RegexSeek(LibrariesString, "(?<=(quiltmc):intermediary:)[^""]*")
+                        If Regex IsNot Nothing Then
+                            _Version.McName = Regex
+                            GoTo VersionSearchFinish
+                        End If
```
原本的正则已经可以匹配了，那把 TODO 的注释删了就 OK

LTCatt commented on Plain Craft Launcher 2/Pages/PageDownload/ModDownloadLib.vb:
```diff
@@ -1877,6 +1877,114 @@ Retry:
+            '构造步骤加载器
+            Dim Loaders As New List(Of LoaderBase)
+            '下载
+            'BMCLAPI 不支持 Quilt Installer 下载
```
前面加个 `'TODO: `

LTCatt commented on Plain Craft Launcher 2/Pages/PageDownload/PageDownloadInstall.xaml.vb:
```diff
@@ -431,6 +485,48 @@
+            Else
+                BtnQSLClear.Visibility = Visibility.Visible
+                ImgQSL.Visibility = Visibility.Visible
+                LabQSL.Text = SelectedQSL.DisplayName.Split("]")(1).Replace("Quilt API ", "").Replace(" build ", ".").Split("+").First.Trim
```
`.Split("+").First` 可以使用 `.Before("+")` 代替

LTCatt replied:
类似地，`.Split("]").Last` 可以使用 `.After("]")` 代替

LTCatt commented on Plain Craft Launcher 2/Pages/PageDownload/PageDownloadInstall.xaml.vb:
```diff
@@ -688,11 +800,11 @@
         If LoadOptiFine.State.LoadingState = MyLoading.MyLoadingState.Error Then Return "获取版本列表失败：" & CType(LoadOptiFine.State, Object).Error.Message
         '检查 Forge 1.13 - 1.14.3：全部不兼容
-        If SelectedForge IsNot Nothing AndAlso
+        If (SelectedLoaderName = "Forge") AndAlso
```
不需要括号，不是 C#……

LTCatt commented on Plain Craft Launcher 2/Pages/PageDownload/PageDownloadInstall.xaml.vb:
```diff
@@ -869,8 +981,7 @@
             If Version.Category = "universal" OrElse Version.Category = "client" Then Continue For '跳过无法自动安装的版本
-            If SelectedNeoForge IsNot Nothing Then Return "与 NeoForge 不兼容"
-            If SelectedFabric IsNot Nothing Then Return "与 Fabric 不兼容"
+            If SelectedLoaderName IsNot Nothing AndAlso SelectedLoaderName IsNot "Forge" Then Return $"与 {SelectedLoaderName} 不兼容"
```
`If SelectedLoaderName <> "Forge" Then`
如果是 null 会返回 false 的，后面的也一样

Pigeon0v0 replied:
不知道为什么改了就炸...

LTCatt commented on Plain Craft Launcher 2/Pages/PageDownload/PageDownloadInstall.xaml.vb:
```diff
@@ -1174,6 +1289,165 @@
+    Private Function LoadQuiltGetError() As String
+        If LoadQuilt Is Nothing OrElse LoadQuilt.State.LoadingState = MyLoading.MyLoadingState.Run Then Return "正在获取版本列表……"
+        If LoadQuilt.State.LoadingState = MyLoading.MyLoadingState.Error Then Return "获取版本列表失败：" & CType(LoadQuilt.State, Object).Error.Message
+        'For Each Version As JObject In DlQuiltListLoader.Output.LoaderValue
+        '    Hint(Version("version").ToString)
+        '    If Version("version").ToString = SelectedMinecraftId.Replace("∞", "infinite").Replace("Combat Test 7c", "1.16_combat-3") Then
+        '        If SelectedForge IsNot Nothing Then Return "与 Forge 不兼容"
+        '        If SelectedNeoForge IsNot Nothing Then Return "与 NeoForge 不兼容"
+        '        Return Nothing
+        '    End If
+        'Next
```
这段可以删了……？

LTCatt commented on Plain Craft Launcher 2/Pages/PageDownload/PageDownloadLeft.xaml.vb:
```diff
@@ -149,7 +157,7 @@
         AniControlEnabled += 1
         If Not Setup.Get("HintHandInstall") Then
             Setup.Set("HintHandInstall", True)
-            If MyMsgBox("手动安装包功能提供了 OptiFine、Forge 等组件的 .jar 安装文件下载，但无法自动安装。" & vbCrLf &
+            If MyMsgBox("手动安装包功能提供了 OptiFine、Forge 等组件的 Jar 安装文件下载，但无法自动安装。" & vbCrLf &
```
就写 .jar 就行

LTCatt commented on Plain Craft Launcher 2/Pages/PageSetup/PageSetupLaunch.xaml:
```diff
@@ -246,7 +246,7 @@
                                         ToolTip="警告：强烈推荐在版本设置中关闭特定版本的警告，而不是在这里关闭所有版本的警告！&#xa;如果在设置中手动选择了与当前版本不兼容的 Java，则自动跳过兼容性警告弹窗，强制使用手动选择的 Java。" />
                         <local:MyCheckBox Text="关闭文件校验" x:Name="CheckAdvanceAssets" Tag="LaunchAdvanceAssets"
-                                        ToolTip="警告：强烈推荐在版本设置中关闭特定版本的校验，而不是在这里关闭所有版本的校验！&#xa;完全不更改 assets；不校验 libraries、第三方登录库与版本主 jar 文件是否被修改。&#xa;如果你没有修改相关文件，请勿勾选此项。" />
+                                        ToolTip="警告：强烈推荐在版本设置中关闭特定版本的校验，而不是在这里关闭所有版本的校验！&#xa;完全不更改 Assets；不校验 Libraries、第三方登录库与版本主 Jar 文件是否被修改。&#xa;如果你没有修改相关文件，请勿勾选此项。" />
```
这里是刻意使用小写以与文件夹名一致的

LTCatt reviewed (changes requested):
对注释、Log 和大小写的修改都没啥必要，这些都是内部写法……
因为现在的 PR 其实都是我一边手动复制到代码里一边重新整理的，所以这些注释修改最后应该也不会合进去 Orz

Pigeon0v0 committed:
refactor: 现在只需要请求一次 API 就可以获取所需信息
revert: Fabric 和 Quilt 获取版本号重新合并到一起
chores: 代码清理

LTCatt commented:
现在堆着的有点多，我看多久动态发个投票来决定先整哪个吧……

allMagicNB commented:
我堆了个 Pre-Release Branch（

Pigeon0v0 committed:
fix: 自动安装选择 1.14 及以前的只有两位版本号的 MC 时意外崩溃
fix: 错误地显示部分快照版支持 Quilt

Pigeon0v0 committed:
chores: 不知道啥时候改了一行测试没改回来

Pigeon0v0 commented on Plain Craft Launcher 2/Pages/PageDownload/PageDownloadInstall.xaml.vb:
```diff
@@ -431,6 +485,48 @@
             End If
         End If
+        'Quilt
+        If SelectedMinecraftId.Contains("1.") AndAlso Val(SelectedMinecraftId.Split(".")(1)) <= 14 AndAlso Val(SelectedMinecraftId.Split(".")(2)) < 4 Then
```
Fixed in b9831cf.

Pigeon0v0 replied:
Fixed in 7a2478b.

Pigeon0v0 committed:
fix: 自动安装选择 1.14 会意外崩溃

3gf8jv4dv reviewed (commented):
Now selecting `1.14` no longer crashes. 👍 

No other issues have been found so far. Waiting for others’ opinions.

3gf8jv4dv reviewed (changes requested):
It seems that if you select `1.14.4` for the first time after opening the launcher, a strange scene will appear in the UI. Or select a version before 1.14.4 and then select `1.14.4` and it will also appear. It will be normal if you select a higher version and then select `1.14.4`.

<details>
<summary>Preview</summary>

![](https://github.com/user-attachments/assets/dd16017e-b24c-4de5-a387-77b7b08a59ad)

</details>

Pigeon0v0 commented:
I don't know why it happens... 

Wait for LTCatt. I have no computer at my hands right now.

Pigeon0v0 commented:
In my opinion, QSL / QFAPI is designed specially for Quilt, and it can replace Fabric API, so I didn't let Fabric API appear after you have selected Quilt.

hejiehao commented:
> In my opinion, QSL / QFAPI is designed specially for Quilt, and it can replace Fabric API, so I didn't let Fabric API appear after you have selected Quilt.

有一些 Minecraft 版本 QFAPI 不支持或者更新速度较慢，这个时候可以用 Fabric API 来替代

Pigeon0v0 commented:
> 有一些 Minecraft 版本 QFAPI 不支持或者更新速度较慢，这个时候可以用 Fabric API 来替代

好的，有空了改（

Pigeon0v0 committed:
Merge branch 'main' into quilt

Pigeon0v0 committed:
feat: 允许 Quilt 安装 Fabric API
fix: 自动安装选择 1.14.4 时 Quilt 卡片可能显示异常

Pigeon0v0 commented:
Done in 6d88559. 

Now Fabric API is able to be selected with Quilt selected. I've also added a hint when you select both Quilt and Fabric API, but QSL is actually available. If QSL is not available, Fabric API will be selected automatically. 

@3gf8jv4dv @hejiehao Have a test and leave comments. :)

<details>
<summary>Preview</summary>

![image](https://github.com/user-attachments/assets/261686f9-301c-47fa-8c42-be24742ccecf)

![image](https://github.com/user-attachments/assets/437c38ed-fec7-48ff-8436-3bb6843c8069)

![image](https://github.com/user-attachments/assets/19355aec-2892-4b5f-abf4-c8d75c64d4d8)

![image](https://github.com/user-attachments/assets/1a11a1ec-dcca-4f29-871f-08f57580ea43)
</details>

Pigeon0v0 committed:
fix: 在选择了 Fabric API / QSL 后直接取消选择 Fabric / Quilt 不会重置已选择的 API 选项

Deep-Dark-Forest commented on Plain Craft Launcher 2/Modules/Base/ModBase.vb:
```diff
@@ -1205,7 +1205,7 @@ Re:
                         GetJson(Content)
                     Catch ex As Exception
-                        Throw New Exception("不是有效的 json 文件", ex)
+                        Throw New Exception("不是有效的 Json 文件", ex)
```
为什么不是 JSON？

Pigeon0v0 replied:
这没必要...

3gf8jv4dv reviewed (commented):
Has a UI issue, otherwise looks good. (Not sure if it can be fixed)

**EDIT:** *This also happens with Fabric Loader.*

<details>
<summary>Preview</summary>

![](https://github.com/user-attachments/assets/043f0625-e8c4-4fe6-9295-a8aed477698c)

</details>

Pigeon0v0 commented:
I don't know how it happens...

JingHai-Lingyun commented:
button event binding issue?

JingHai-Lingyun commented:
I this #1784 need another pr. 
Because if it can detector mods in folder, ui may has changes.

Pigeon0v0 committed:
Merge branch 'main' into quilt

Pigeon0v0 committed:
fix: 不知道哪里冒出来的 Exit Sub 和不知道为什么没有同步上游修改导致的重复任务检查爆炸

allMagicNB commented on Plain Craft Launcher 2/Pages/PageDownload/ModDownloadLib.vb:
```diff
@@ -310,6 +310,47 @@ Public Module ModDownloadLib
+        Try
+            Dim Id As String = DownloadInfo.NameVersion
+            Dim VersionFolder As String = PathMcFolder & "versions\" & Id & "\"
+            Dim IsNewVersion As Boolean = Val(DownloadInfo.Inherit.Split(".")(1)) >= 14
+            Dim Target As String = If(IsNewVersion,
+                PathTemp & "Cache\Code\" & DownloadInfo.NameVersion & "_" & GetUuid(),
+                PathMcFolder & "libraries\optifine\OptiFine\" & DownloadInfo.NameFile.Replace("OptiFine_", "").Replace(".jar", "").Replace("preview_", "") & "\" & DownloadInfo.NameFile.Replace("OptiFine_", "OptiFine-").Replace("preview_", ""))
+
+            '重复任务检查
+            SyncLock LoaderTaskbarLock
+                For i = 0 To LoaderTaskbar.Count - 1
+                    If LoaderTaskbar(i).Name = "OptiFine " & DownloadInfo.NameDisplay & " 下载" Then
+                        Hint("该版本正在下载中！", HintType.Critical)
+                        Exit Sub
+                    End If
+                Next
+            End SyncLock
+
+            '已有版本检查
+            If File.Exists(VersionFolder & Id & ".json") Then
+                If MyMsgBox("版本 " & Id & " 已存在，是否重新下载？" & vbCrLf & "这会覆盖版本的 Json 和 Jar 文件，但不会影响版本隔离的文件。", "版本已存在", "继续", "取消") = 1 Then
+                    File.Delete(VersionFolder & Id & ".jar")
+                    File.Delete(VersionFolder & Id & ".json")
+                Else
+                    Exit Sub
+                End If
+            End If
+
+            '启动
+            Dim Loader As New LoaderCombo(Of String)("OptiFine " & DownloadInfo.NameDisplay & " 下载", McDownloadOptiFineLoader(DownloadInfo)) With {.OnStateChanged = AddressOf McInstallState}
+            Loader.Start(VersionFolder)
+            LoaderTaskbarAdd(Loader)
+            FrmMain.BtnExtraDownload.ShowRefresh()
+            FrmMain.BtnExtraDownload.Ribble()
+
+        Catch ex As Exception
+            Log(ex, "开始 OptiFine 下载失败", LogLevel.Feedback)
+        End Try
+    End Sub
+
```
？
